### ISSUE TITLE
Include a way to know what models are referenced by current model (similar to References tab in RSA)

### ISSUE THREAD

**Issue by J17359** [ZeligsoftDev/CX4CBDDS#270]
# **Issue and tracking information**
### Developer's Actual time spent on fix (hours)  
# **Issue reporter to provide a detailed description of the issue in the space below**
Need a way to know (similar to RSA's References tab) what models reference and what models are referenced by the current model. It would also be good to know (RSA didn't have this feature) what the path to the model files is, so we know if it's hooked to the correct copy of the model in the case of dynamic path maps.

**Comment by J17359** [comment]
@ysroh - Before jumping into fixing this, please throw out some ideas for how to solve this. We are really interested in dynamic path mapping and really want to ensure that it is usable and errors/warnings are friendly and instructive for users that have missing models/broken links in their workspace.

**Comment by ysroh** [comment]
@J17359 There are some checks done already regarding the safeguard of Dynamic Pathmap usage. See Issue #191 and Issue #242 for details.

Here are available safeguards,

1. Upon opening a new model following checks are done in order to ensure that all required models are available in the workspace.
   - Check all package imports to see if any imported package is missing and warn the user.
![image](https://user-images.githubusercontent.com/896503/126514102-9714cfa0-eff5-4c33-92fa-ebdaf14d5ede.png)
   - Check all model elements' references to see if any dynamic pathmap target is missing and warn the user.
![image](https://user-images.githubusercontent.com/896503/126514921-bb1e6eaf-0aae-454d-9ecd-07692ea56bae.png)
   - You see the details of missing references when the model is validated
![image](https://user-images.githubusercontent.com/896503/126515046-a93a76a2-c2e0-47e2-bd17-cebd0129af6d.png)

2. Upon deletion of a dynamic pathmap model all models in the workspace are checked to see if there exist any references to the model deleted and notify users of dependent models. Users can either bring the deleted model back to the workspace or open the dependent models to fix broken references.
![image](https://user-images.githubusercontent.com/896503/126516390-b5c03e0a-f52b-425e-9ba9-cc9053bf932f.png)

There is no references tab that shows all the dependent models like in RSA but you can see all the reference models when you open the model without editing the model.  See snapshot below.
![image](https://user-images.githubusercontent.com/896503/126649722-581e7694-194d-4a3c-8048-d3b6a69d27b8.png)

Let us know if these safeguards would be sufficient.

**Comment by j26151** [comment]
@ysroh I think the feature as you implemented it works really well for Dynamic pathmapping. One thing that we want to address here as well is not dynamic pathmapped ICMs. This is the way we are doing it in RSA. We obviously hope to one day move all of our modelling to using dynamic pathmapping, but until then, we want to make sure that the old way works as well. 

If I take two example models, say ClientServer and BasicPubSub and have them in the same workspace. Then, I go and type the return value in the interface operation in the ClientServer's ports folder with a type from BasicPubSub (using the browse function within the type dialog and not using dynamic pathmapping), two things happen that we do not want.

1. The model explorer now shows every single model in the workspace, which I assume is because papyrus thinks we may want to pull references from everywhere (which is fair), but this now looks like our ClientServer model includes every model.
2. If we close ClientServer and re-open it, only the models ClientServer has imported are shown. BPS is not shown (and it should be since we are using a type from it), but if I navigate down to the operation where we used the type for BPS, the BPS model pops up like it should have been all along. Here is a before and after of the same model. All that is different is that I have navigated to the spot in ClientServer where the outside type is referenced (also, i used PubSubAttachment and another internal model CCM_PSAT for testing this. But BPS would show the same):
![image](https://user-images.githubusercontent.com/49879646/126874962-c34368a3-a1d8-47e9-8792-7fc87ebb67fa.png)
![image](https://user-images.githubusercontent.com/49879646/126874968-a072a6eb-fdd9-431f-a8cc-367c2fcb1ad8.png)

**Comment by ysroh** [comment]
@j26151 You are correct. The referencing models are only correct when the model is open so they do not always reflect the correct model list. They just reflect all model resources that are loaded in the current resource set. You can load any model into your current model so you can make a reference to an element in that model.

What if we check all broken references, not just Pathmap references? when the model is open and deleted? So we give similar error messages that we do for broken pathmap references?

**Comment by j26151** [comment]
@ysroh your suggestion about checking all broken references is good.

However, I am still concerned about the bug the images show. The only way that I would know that the ClientServer model includes PubSubAttachment and CCM_PSAT, is if i not only open the model, but open the model and navigate all the way down the very specific place that those models are referenced.

Additionally, the bug mentioned in my first bullet above should be addressed.

**Comment by ysroh** [comment]
@j26151 

> @ysroh your suggestion about checking all broken references is good.
> 
> However, I am still concerned about the bug the images show. The only way that I would know that the ClientServer model includes PubSubAttachment and CCM_PSAT, is if i not only open the model, but open the model and navigate all the way down the very specific place that those models are referenced.

ys] I will fix this so all referenced models are shown when the model is open

> 
> Additionally, the bug mentioned in my first bullet above should be addressed.
ys] This is how Papyrus work if you want to find an element that does not belong to the current model from the selection dialog. In order to make this work, I can load all resources from the model and unload all resources that are not used once the selection is made. I can do this but we need to load and unload resources repeatedly whenever the selection dialog is open.

The other way we could do this is the following.
We can remove options from the selection dialog.
![image](https://user-images.githubusercontent.com/896503/127016901-ff8c5cc4-b400-4c3f-aecf-748d2ac444ff.png)
The search scope would be all elements from your model and from the package imported models.
If you want to reference a model in your workspace, then you need to create a package import of that model through the import menu provided. There is a menu to import ICM models so I can add another menu to create non-ICM models in the workspace.

I think either way would work. Can you let me know which way you would prefer?
If you like the flexibility of selecting an element from any model without importing them then the first option should be chosen. I just need to make sure any unused references are unloaded from the model set.
Let me know.

**Comment by j26151** [comment]
I believe before when I did this, I used the browse tab at the top, mainly because I forgot about those options at the bottom.  I think we want to remove the "entire workspace" option, as SNA does not "approve of" using models with references in the workspace unless they are dynamically pathmapped.

So I think we want a hybrid approach with the following two fixes for this problem:

1. Remove the "Search Entire Workspace" checkbox
2. Unload all resources that are not being used after the selection is made

**Comment by ysroh** [comment]
> I believe before when I did this, I used the browse tab at the top, mainly because I forgot about those options at the bottom. I think we want to remove the "entire workspace" option, as SNA does not "approve of" using models with references in the workspace unless they are dynamically pathmapped.

> 
> So I think we want a hybrid approach with the following two fixes for this problem:
> 
> 1. Remove the "Search Entire Workspace" checkbox

ys] I just want to confirm that you still want to allow users to make a reference to another model within the same project as the model you are editing? even if the model is not pathmapped? 

> 2. Unload all resources that are not being used after the selection is made

**Comment by j26151** [comment]
Yes. When models are dynamically pathmapped (in the workspace or project scope), they have to first be imported as an ICM, which makes their types all able to be referenced and there are no issues.

However, if they are not using dynamically pathmapped models, we do not want them to be able to reference models outside of the project. However, in RSA, we currently do allow them to reference all models within the project. So, yes, we want the users to be able to make references to another model within the same project as the model we are not editing, even if the model is not pathmapped.

**Comment by ysroh** [comment]
I made some changes to delete the "search entire workspace" option and tried to unload models that are not referenced, but I realized that it is much complicated than I thought. In fact, it is almost impossible to keep the reference list clean except when the model is open.

I will come up with another solution for this. 
I am thinking of two approaches right now. One way is to add a menu to pop up a new dialog to show the reference list and another one is to use the properties view of the Model element to show the reference list at the bottom of the Properties view. 

I think the properties view option would work well but I will let you know after I try it.

**Comment by ysroh** [comment]
I've committed my fix to "papyrus_bugfix" branch. I haven't merged this branch to papyrus branch yet because I wanted to make sure that this solution works for your team first.

As I suggested in the comment above, the new "referenced model" section is added to the properties view of the top model element. You need to click on the "Show Referenced Models" button to calculate references and show them in the table because it is too expensive to automatically calculate references each time when the model is clicked. The list in the table can be expanded to show all referencing elements. Any referenced model that cannot be found is marked with the error icon so users can easily see which models are missing.

![image](https://user-images.githubusercontent.com/896503/128106721-02080417-4d06-4792-8a5a-f30a755ce1dd.png)

It would be great if you could give it a try and let me know.

**Comment by ysroh** [comment]
I am also putting a tab for other models that are referencing the current model.

![image](https://user-images.githubusercontent.com/896503/128379079-d3cfc805-b362-41d9-b12e-d902c4583f91.png)

**Comment by ysroh** [comment]
Merged fix to papyrus branch. Let me know if I can close this issue. Thanks

**Comment by j26151** [comment]
This looks great, thank you. I think we can close this issue now.
